[CORE] Add priority to Zone

tomkalon opened the pull request:
| Q               | A
|-----------------|-----
| Branch?         | 2.2
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| License         | MIT

The following PR fixes have been adapted to Sylius 2.2

https://github.com/Sylius/Sylius/pull/17191

> The PR https://github.com/Sylius/Sylius/pull/15275 broke the tax calculation in our shop. The reason being that the ZoneMatcher now returns the first matching zone, while previously it was the last.

Commit:
[Zone] Added position property

(cherry picked from commit 686daab0a3c8d64a70eeae0cfab34b907233f6cb)

tomkalon committed:
Rename `position` to `priority` across the Zone model and associated components.

tomkalon committed:
Add Behat tests for sorting zones by priority

tomkalon committed:
Add priority column to AddressingBundle SQLlite db and update related test expectations.

NoResponseMate commented on features/admin/addressing/managing_zones/editing_zone.feature:
```diff
@@ -67,3 +67,13 @@ Feature: Editing a zone
+        And I set its priority to "-1"
+        And I save my changes
+        Then I should be notified that it has been successfully edited
+        And the "European Union" zone should have priority 2
```
```suggestion
    Scenario: Setting zone to the highest priority
        Given the store has a zone "European Union" with code "EU" and priority 0
        And it has the "France" country member
        When I want to modify the zone named "European Union"
        And I set its priority to "-1"
        And I save my changes
        Then I should be notified that it has been successfully edited
        And the "European Union" zone should have priority 2
```

tomkalon replied:
👍🏻  fixed

NoResponseMate commented on src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/Zone.orm.xml:
```diff
@@ -23,6 +24,10 @@
 
+        <field name="priority" type="integer">
+            <gedmo:sortable-position />
+        </field>
```
Do we really need gedmo for this?
It'll force shuffling all priorities when we add a new zone with prio 0, also having duplicated priorities doesn't seem like a problem since they already are separate by scope. With this change, priority becomes basically unique throughout the whole zone table.

tomkalon replied:
👍🏻  fixed

NoResponseMate commented on src/Sylius/Bundle/CoreBundle/Migrations/Version20241010135921.php:
```diff
@@ -0,0 +1,27 @@
+<?php
+
+declare(strict_types=1);
```
Missing license, same below

tomkalon replied:
👍🏻  fixed

NoResponseMate commented on src/Sylius/Behat/Element/Admin/Zone/FormElement.php:
```diff
@@ -67,6 +72,12 @@ public function addMember(): void
+    {
+        $this->getElement('priority')->setValue($priority);
+    }
+
```
```suggestion
```

tomkalon replied:
👍🏻  fixed

tomkalon committed:
Update migration files with license headers and adjust zone priority handling

- Added Sylius license headers to migration files.
- Removed `gedmo:sortable-position` from `priority` field in Zone ORM mapping.
- Updated Behat scenario to reflect changes in zone priority expectations.

NoResponseMate commented on src/Sylius/Bundle/AddressingBundle/Resources/config/doctrine/model/Zone.orm.xml:
```diff
@@ -11,7 +11,8 @@
 
-<doctrine-mapping xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping">
+<doctrine-mapping xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping"
+                  xmlns:gedmo="http://gediminasm.org/schemas/orm/doctrine-extensions-mapping">
```
```suggestion
<doctrine-mapping xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping">
```

NoResponseMate commented on src/Sylius/Bundle/CoreBundle/Migrations/Version20241010135921.php:
```diff
@@ -0,0 +1,36 @@
+    public function up(Schema $schema): void
+    {
+        $this->addSql('ALTER TABLE sylius_zone ADD priority INT NOT NULL DEFAULT 0');
+        $this->addSql('UPDATE sylius_zone SET priority = id');
```
```suggestion
```
Wouldn't this flip the currently used order?

tomkalon committed:
Remove redundant `gedmo` namespace and problematic SQL update statement
